relay-metrics/src/aggregation.rs

```diff
@@ -1259,6 +1310,38 @@ impl Aggregator {
 
         let key = Self::validate_bucket_key(key, &self.config)?;
 
+        // XXX: This is not a great implementation of cost enforcement.
+        //
+        // * it takes two lookups of the project key in the cost tracker to merge a bucket: once in
+        //   `check_limits_exceeded` and once in `add_cost`.
```

Comment:
Could you try to return the error from `add_cost` and do the diff computation inline? The `.entry()` function below doesn't create the entry until you call `insert` on it at least, which just leaves the merge into the bucket, and that could also easily return whether it did insert or not. 

It's excellent that you're calling out caveats of enforcement below, though in practice an off-by-one is not an issue.